server/kv.go

```diff
@@ -0,0 +1,694 @@
+	for _, instanceID := range instances.IDs() {
+		_, err = p.userStore.LoadConnection(instanceID, mattermostUserID)
+		if errors.Cause(err) == kvstore.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		user.ConnectedInstances.Set(instances.Get(instanceID))
+	}
+	err = p.userStore.StoreUser(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return user, nil
+}
+
+func (store store) StoreOAuth2State(state string) error {
+	// Expire in 15 minutes
+	appErr := store.plugin.API.KVSetWithExpiry(
+		hashkey(prefixOneTimeSecret, state), []byte(state), 15*60)
+	if appErr != nil {
+		return errors.WithMessage(appErr, "failed to store state "+state)
+	}
+	return nil
+}
+
+func (store store) VerifyOAuth2State(state string) error {
+	data, appErr := store.plugin.API.KVGet(hashkey(prefixOneTimeSecret, state))
+	if appErr != nil {
+		return errors.WithMessage(appErr, "failed to load state "+state)
+	}
+
+	if string(data) != state {
+		return errors.New("invalid oauth state, please try again")
+	}
+
+	return nil
+}
+
+func (store store) EnsureAuthTokenEncryptionSecret() (secret []byte, returnErr error) {
+	defer func() {
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr, "failed to ensure auth token secret")
+	}()
+
+	// nil, nil == NOT_FOUND, if we don't already have a key, try to generate one.
+	secret, appErr := store.plugin.API.KVGet(keyTokenSecret)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	if len(secret) == 0 {
+		newSecret := make([]byte, 32)
+		_, err := rand.Reader.Read(newSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		appErr = store.plugin.API.KVSet(keyTokenSecret, newSecret)
+		if appErr != nil {
+			return nil, appErr
+		}
+		secret = newSecret
+		store.plugin.debugf("Stored: auth token secret")
+	}
+
+	// If we weren't able to save a new key above, another server must have beat us to it. Get the
+	// key from the database, and if that fails, error out.
+	if secret == nil {
+		secret, appErr = store.plugin.API.KVGet(keyTokenSecret)
+		if appErr != nil {
+			return nil, appErr
+		}
+	}
+
+	return secret, nil
+}
+
+func (store store) EnsureRSAKey() (rsaKey *rsa.PrivateKey, returnErr error) {
+	defer func() {
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr, "failed to ensure RSA key")
+	}()
+
+	err := store.get(keyRSAKey, &rsaKey)
+	if err != nil && errors.Cause(err) != kvstore.ErrNotFound {
+		return nil, err
+	}
+
+	if rsaKey == nil {
+		var newRSAKey *rsa.PrivateKey
+		newRSAKey, err = rsa.GenerateKey(rand.Reader, 1024) // #nosec G403
+		if err != nil {
+			return nil, err
+		}
+
+		err = store.set(keyRSAKey, newRSAKey)
+		if err != nil {
+			return nil, err
+		}
+		rsaKey = newRSAKey
+		store.plugin.debugf("Stored: RSA key")
+	}
+
+	// If we weren't able to save a new key above, another server must have beat us to it. Get the
+	// key from the database, and if that fails, error out.
+	if rsaKey == nil {
```

Comment:
This code is repeated above, why not creating a method so we DRY?

Reply:
we checking this below like the first condition is if the error is not nil then return the error then we are checking whether the len(secret) == 0 then we are generating a new secret for that so here we won't be returning the error is both are nil.